common/src/main/java/io/netty/util/internal/MpscLinkedQueue.java

```diff
@@ -248,53 +256,37 @@ public E element() {
         throw new NoSuchElementException();
     }
 
-    @Override
-    public Object[] toArray() {
-        final Object[] array = new Object[size()];
-        final Iterator<E> it = iterator();
-        for (int i = 0; i < array.length; i ++) {
-            if (it.hasNext()) {
-                array[i] = it.next();
-            } else {
-                return Arrays.copyOf(array, i);
+    private List<E> toList(int initialCapacity) {
+        List<E> elements = new ArrayList<E>(initialCapacity);
+
+        MpscLinkedQueueNode<E> n = peekNode();
+        for (;;) {
+            if (n == null) {
+                break;
             }
+            E value = n.value();
+            if (value == null) {
+                break;
+            }
+            elements.add(value);
+            MpscLinkedQueueNode<E> next = n.next();
+            if (n == next) {
+                break;
+            }
+            n = next;
         }
-        return array;
+        return elements;
+    }
+
+    @Override
+    public Object[] toArray() {
+        return toList(16).toArray();
```

Comment:
Does the runtime type of `toList(..)` have to be `ArrayList<>`?  Can you just change the signature of  `toList(..)` to accept a `List<..>` since it is private?  That way we don't have to duplicate the `16 is the default size` logic here.